### DISCUSSION TITLE
Renovate doesn't open PR's for certain vulnerabilityAlerts (Case sensitive packages?)

### DISCUSSION THREAD

**Discussion by jeroenmaas** [renovatebot/renovate#29230] [Request Help] [answered]
### What would you like help with?

I think I found a bug

### How are you running Renovate?

Mend Renovate hosted app on github.com

### Please tell us more about your question or problem

We're running renovate with vulnerabilityAlerts enabled. Github shows a few vulnerabilities and with our current configuration a few nice PR's have been made! But the stream of PR's has been resolved, no new ones are created but there are still a few vulnerabilities  shown on Github. 

There are 2 effected packages:

1. Django (==4.1.7 for testing purposes)
2. Requests (==2.31.0)

Digging around for Django I believe I might know the cause. https://github.com/advisories/GHSA-vm8q-m57g-pff3 is writted as django (lowercase d) and no PR is created. https://github.com/advisories/GHSA-qmf9-6jqf-j8fq is for the package Django and is applied. Renovate targets "Django" if you want to use matchPackageNames.

Why requests doesn't work is a mystery to me. Logs would suggest it thinks 2.31.0 already has the bug fixed.

I've looked for similar issues and https://github.com/renovatebot/renovate/discussions/27069 seems to be the same. But there has been a patch for that and i'm using a later version of renovate. ("37.368.10" was used though it hosted app on github.com)

Configuration:
```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:recommended"
  ],
  "packageRules": [
    {
      "enabled": false,
      "description": "Group all dependencies from frontend projects",
      "matchFileNames": ["fontend/**"],
      "groupName": "Frontend dependencies"
    },
    {
      "enabled": false,
      "description": "Group all dependencies from backend projects",
      "matchFileNames": ["backend/**"],
      "groupName": "Backend dependencies"
    },
    {
      "enabled": false,
      "description": "Github / docker deps",
      "matchFileNames": [".github/**", "docker/**", "docker-compose.yml", "docker-compose.prod.yml"],
      "groupName": "Misc dependencies"
    }
  ],
  "vulnerabilityAlerts": {
    "enabled": true
  },
  "pip_requirements": {
    "fileMatch": ["(^|/)[\\w-]*base([-.]\\w+)?\\.(txt)$"]
  }
}
```

### Logs (if relevant)

<details><summary>Logs</summary>

```

DEBUG: GitHub vulnerability details
{
  "alerts": {
    "pip/django": {
      ">= 4.2, < 4.2.10": "4.2.10",
      ">= 4.2, < 4.2.11": "4.2.11"
    },
    "pip/requests": {
      "< 2.32.0": "2.32.0"
    }
  }
}

DEBUG: alert package rules
{
  "alertPackageRules": [
    {
      "matchDatasources": [
        "pypi"
      ],
      "matchPackageNames": [
        "django"
      ],
      "matchCurrentVersion": "== 4.2.7",
      "matchFileNames": [
        "backend/requirements/base.txt"
      ],
      "allowedVersions": "==4.2.11",
      "prBodyNotes": [
        "### GitHub Vulnerability Alerts",
        "#### [CVE-2024-24680](https://nvd.nist.gov/vuln/detail/CVE-2024-24680)\n\nAn issue was discovered in Django 3.2 before 3.2.24, 4.2 before 4.2.10, and Django 5.0 before 5.0.2. The intcomma template filter was subject to a potential denial-of-service attack when used with very long strings.",
        "#### [CVE-2024-27351](https://nvd.nist.gov/vuln/detail/CVE-2024-27351)\n\nIn Django 3.2 before 3.2.25, 4.2 before 4.2.11, and 5.0 before 5.0.3, the django.utils.text.Truncator.words() method (with html=True) and the truncatewords_html template filter are subject to a potential regular expression denial-of-service attack via a crafted string. NOTE: this issue exists because of an incomplete fix for CVE-2019-14232 and CVE-2023-43665."
      ],
      "isVulnerabilityAlert": true,
      "force": {
        "groupName": null,
        "schedule": [],
        "dependencyDashboardApproval": false,
        "minimumReleaseAge": null,
        "rangeStrategy": "update-lockfile",
        "commitMessageSuffix": "[SECURITY]",
        "branchTopic": "{{{datasource}}}-{{{depName}}}-vulnerability",
        "prCreation": "immediate",
        "enabled": true
      }
    },
    {
      "matchDatasources": [
        "pypi"
      ],
      "matchPackageNames": [
        "requests"
      ],
      "matchCurrentVersion": "== 2.31.0",
      "matchFileNames": [
        "backend/requirements/base.txt"
      ],
      "allowedVersions": "==2.32.0",
      "prBodyNotes": [
        "### GitHub Vulnerability Alerts",
        "#### [CVE-2024-35195](https://github.com/psf/requests/security/advisories/GHSA-9wx4-h78v-vm56)\n\nWhen making requests through a Requests `Session`, if the first request is made with `verify=False` to disable cert verification, all subsequent requests to the same origin will continue to ignore cert verification regardless of changes to the value of `verify`. This behavior will continue for the lifecycle of the connection in the connection pool.\n\n### Remediation\nAny of these options can be used to remediate the current issue, we highly recommend upgrading as the preferred mitigation.\n\n* Upgrade to `requests>=2.32.0`.\n* For `requests<2.32.0`, avoid setting `verify=False` for the first request to a host while using a Requests Session.\n* For `requests<2.32.0`, call `close()` on `Session` objects to clear existing connections if `verify=False` is used.\n\n### Related Links\n* https://github.com/psf/requests/pull/6655"
      ],
      "isVulnerabilityAlert": true,
      "force": {
        "groupName": null,
        "schedule": "[Circular]",
        "dependencyDashboardApproval": false,
        "minimumReleaseAge": null,
        "rangeStrategy": "update-lockfile",
        "commitMessageSuffix": "[SECURITY]",
        "branchTopic": "{{{datasource}}}-{{{depName}}}-vulnerability",
        "prCreation": "immediate",
        "enabled": true
      }
    }
  ]
}



{
            "depName": "requests",
            "currentValue": "==2.31.0",
            "datasource": "pypi",
            "currentVersion": "2.31.0",
            "updates": [],
            "packageName": "requests",
            "versioning": "pep440",
            "warnings": [],
            "sourceUrl": "https://github.com/psf/requests",
            "registryUrl": "https://pypi.org/pypi",
            "homepage": "https://requests.readthedocs.io",
            "changelogUrl": "https://github.com/psf/requests/blob/master/HISTORY.md",
            "currentVersionTimestamp": "2023-05-22T15:12:42.000Z",
            "fixedVersion": "2.31.0"
          },

          {
            "depName": "Django",
            "currentValue": "==4.2.7",
            "datasource": "pypi",
            "currentVersion": "4.2.7",
            "updates": [],
            "packageName": "Django",
            "skipReason": "disabled"
          },


```

</details>

**Comment by Churro** [comment]
Needs a reproduction repo, especially to verify the case with `requests`.

For Django, the solution could be to change the alert package rule to match Django case-insensitive with `matchPackagePatterns` instead of `matchPackageNames` but that's up for verification on a reproduction repo.

**Comment by jeroenmaas** [comment]
Thanks for taking the time to look at the request. I've managed to reproduce both issues in this repo: https://github.com/jeroenmaas/renovate-pip-vulnerability-alerts-bug

In order to reproduce Django I had to first merge 4.2.7 PR by first ignoring the other django security items. Otherwise going from ==4.2 would recommend me 4.2.11. I think it might be case insensitive for selecting which version but not case insensitive for bypassing the "enabled" in packageRules

I've added cryptography just to showcase other security items are handled fine and there is no issue with renovate setup.

I've added the logs for additional debug information. If you need any more information please let me know. [log-final-execution-renovate.txt](https://github.com/renovatebot/renovate/files/15422751/log-final-execution-renovate.txt)

**Reply by Churro** [reply, marked as answer]
Thanks for the reproduction repo, I've forked it to https://github.com/renovate-demo/29230-renovate-pip-vulnerability-alerts-bug and ran some tests:

- You see no PR for https://github.com/psf/requests/security/advisories/GHSA-9wx4-h78v-vm56 affecting `requests` due to the fact that the advisory suggests version 2.32.0 as a fix. This version was retracted (see [here](https://pypi.org/project/requests/2.32.0/)). When renovate fetches the list of available requests releases, it filters this version, which eventually results in a situation as if there is no fix available.
- `django` or `Django` should make no difference since PR https://github.com/renovatebot/renovate/pull/28214. However, I spotted in your dependency dashboard that django is longer identified as a dependency (see [here](https://github.com/jeroenmaas/renovate-pip-vulnerability-alerts-bug/issues/2)) after merging https://github.com/jeroenmaas/renovate-pip-vulnerability-alerts-bug/pull/3. That looks bizarre, considering that it is correctly detected in my fork (see [here](https://github.com/renovate-demo/29230-renovate-pip-vulnerability-alerts-bug/issues/3)). My best guess is that this is due a caching issue in renovate. Unfortunately, the logs you have attached don't provide any additional clues what's going on.

**Reply by jeroenmaas** [reply]
Hi Churro,

Thank you for taking the time to test it out. Much appreciated. `requests` is obvious now. It might be nice if it would go to a newer version but I understand that this could be difficult to implement and also be unexpected for some.

For `django` its weird that you do get the PR. I've continued adding renovate to more of our repo's and seen the same behavior where are non related commit could suddenly trigger new PR's to spawn. I've added a commit on the repo and now the `django` pr is created. I think it might be caused because it doesn't like updating the same dependency again in the next commit. But that is just speculation.

The causes I've started with have been resolved. Though the issue that renovate doesn't always open a PR for security items comes back more often for other repo's. I've added a script that manually screams at me for security items that have no PR so I do those manually. But added 2 more example repo's if you want to go hunting to make it more stable so others don't have to.

https://github.com/jeroenmaas/renovate-vulnerability-alerts-not-detected is a PHP/laravel project with some dependencies. `phpseclib` contains no PR but vulnerabilities. The suggested version `2.0.47` exists and has not been pulled like `requests`. Dependabot is able to create a PR manually.

https://github.com/jeroenmaas/renovate-vulnerability-alerts-not-detected-2 is a nodejs package. `graphql` has a vulnerability and should be upgraded to `16.8.1`. No PR is created by renovate but Dependabot is able to create a PR manually. The version `16.8.1` is not pulled.

I've attached the logs from both new repo's.

[renovate-vulnerability-alerts-not-detected-log.txt](https://github.com/renovatebot/renovate/files/15447002/renovate-vulnerability-alerts-not-detected-log.txt)
[renovate-vulnerability-alerts-not-detected2-log.txt](https://github.com/renovatebot/renovate/files/15447001/renovate-vulnerability-alerts-not-detected2-log.txt)

**Reply by Churro** [reply]
Regarding django it might really have been a glitch unrelated to vulnerabilities:
After your dummy commit https://github.com/jeroenmaas/renovate-pip-vulnerability-alerts-bug/commit/31b786d868e0830afe018521acb40e3d049a355c renovate almost instantly updated the [dependency dashboard](https://github.com/jeroenmaas/renovate-pip-vulnerability-alerts-bug/issues/2) and re-added django (click on _edited_ to see the difference).

Regarding `graqhql` and `phpseclib`, the reason you're not getting PRs is that these are transitive dependencies. Most managers in renovate don't extract such indirect dependencies. They are still covered by enabling regular lockfile maintenance.
